DSharpPlus/Cache/DiscordMemoryCache.cs

```diff
@@ -0,0 +1,109 @@
+﻿namespace DSharpPlus.Cache;
+
+using System;
+using System.Threading.Tasks;
+using Entities;
+using Microsoft.Extensions.Caching.Memory;
+
+public class DiscordMemoryCache : IDiscordCache
+{
+    private MemoryCacheEntryOptions _guildOptions;
+    private MemoryCacheEntryOptions _channelOptions;
+    private MemoryCacheEntryOptions _threadOptions;
+    private MemoryCacheEntryOptions _messageOptions;
+    private MemoryCacheEntryOptions _memberOptions;
+    private MemoryCacheEntryOptions _userOptions;
+
+    private MemoryCache _cache = new(new MemoryCacheOptions());
+
+    public DiscordMemoryCache(CacheConfiguration configuration)
+    {
+        ArgumentNullException.ThrowIfNull(configuration);
+        
+        this._guildOptions = configuration.Guild;
+        this._channelOptions = configuration.Channel;
+        this._threadOptions = configuration.Thread;
+        this._messageOptions = configuration.Message;
+        this._userOptions = configuration.User;
+        this._memberOptions = configuration.Member;
+    }
+
+    public ValueTask Add<T>(T entity)
+    {
+        ArgumentNullException.ThrowIfNull(entity);
+
+        switch (entity)
+        {
+            case DiscordGuild guild:
+                this._cache.Set(new GuildCacheKey(guild.Id), guild, this._guildOptions);
+                break;
+            case DiscordThreadChannel thread:
+                this._cache.Set(new ChannelCacheKey(thread.Id), thread, this._threadOptions);
+                break;
+            case DiscordChannel channel:
+                this._cache.Set(new ChannelCacheKey(channel.Id), channel, this._channelOptions);
+                break;
+            case DiscordMessage message:
+                this._cache.Set(new MessageCacheKey(message.Id), message, this._messageOptions);
+                break;
+            case DiscordMember member:
+                this._cache.Set(new MemberCacheKey(member.Id, member._guild_id), member, this._userOptions);
+                break;
+            case DiscordUser user:
+                this._cache.Set(new UserCacheKey(user.Id), user, this._userOptions);
+                break;
+            default:
+                throw new ArgumentException(
+                    "Invalid type provided. Only DiscordUser, DiscordGuild, DiscordChannel, DiscordThread, and DiscordMessage are supported.");
```

Comment:
Does this correctly handle `DiscordForumChannel`?

Reply:
it should™. Threads are only seperate because we "cached" them seperate before. But i think it would be more robust to use `typeof(entity).IsAssignableFrom(DiscordChannel)`

Reply:
Yes, but the forum channel is a separate channel type that also derives from `DiscordChannel`; pattern matching types can be pretty finicky, hence why I'm asking for confirmation, as this could explode at runtime unexpectedly.

Reply:
I will test it when we integrate it

Reply:
personally, i think we could break the separation of threads and other channels in caching, but this in theory this is fine once tested and working